core/plugins/XConeProducer.cc

```diff
@@ -157,122 +169,99 @@ XConeProducer::produce(edm::Event& iEvent, const edm::EventSetup& iSetup)
-  vector<PseudoJet> particle_in_fat1, particle_in_fat2;
 
-  // get one set of particles for each jet
-  for (unsigned int ipart=0; ipart < _psj.size(); ++ipart) {
-    if (list_fat[ipart]==0) {
-      particle_in_fat1.push_back(_psj.at(ipart)); // get vector as PseudoJet for fatjet 1
-    }
-    if (list_fat[ipart]==1) {
-      particle_in_fat2.push_back(_psj.at(ipart)); // get vector as PseudoJet for fatjet 2
+  // loop over all fatjets and cluster subjets
+  for(unsigned int i=0; i<fatjets.size(); i++){
+    // get set of particles in fatjet i
+    vector<PseudoJet> particle_in_fatjet;
+    for (unsigned int ipart=0; ipart < _psj.size(); ++ipart) {
+      if (list_fat[ipart] < 0) continue;
+      else if (abs(list_fat[ipart])==i) {
+        particle_in_fatjet.push_back(_psj.at(ipart));
+      }
     }
-  }
-
-  uint minParticlesPerJet = 3;
-  bool doJet1Subjets = (particle_in_fat1.size() > minParticlesPerJet);
-  if (!doJet1Subjets) {
-    edm::LogWarning("InsufficientParticles")
-      << "Not enough particles in fatjet 1 to run second XCone step: "
-      << std::to_string(particle_in_fat1.size())
-      << std::endl;
-  }
-
-  bool doJet2Subjets = (particle_in_fat2.size() > minParticlesPerJet);
-  if (!doJet2Subjets) {
-    edm::LogWarning("InsufficientParticles")
-      << "Not enough particles in fatjet 2 to run second XCone step: "
-      << std::to_string(particle_in_fat2.size())
-      << std::endl;
-  }
 
-  // Run second clustering step (N=3, R=0.4) for each fat jet
-  vector<PseudoJet> subjets_1, subjets_2;
-
-  // subjets from fat jet 1
-  vector<double> subjet1_area;
-  if (doJet1Subjets) {
-    std::unique_ptr<NjettinessPlugin> plugin_xcone_sub1;
-    initPlugin(plugin_xcone_sub1, 3, 0.4, 2.0, usePseudoXCone_);
-    JetDefinition jet_def_sub1(plugin_xcone_sub1.get());
-    ClusterSequenceArea clust_seq_sub1(particle_in_fat1, jet_def_sub1, area_def);
-    subjets_1 = sorted_by_pt(clust_seq_sub1.inclusive_jets(0));
-    for (unsigned int j = 0; j < subjets_1.size(); ++j) subjet1_area.push_back(subjets_1[j].area());
-  }
-
-  //subjets from fat jet 2
-  vector<double> subjet2_area;
-  if (doJet2Subjets) {
-    std::unique_ptr<NjettinessPlugin> plugin_xcone_sub2;
-    initPlugin(plugin_xcone_sub2, 3, 0.4, 2.0, usePseudoXCone_);
-    JetDefinition jet_def_sub2(plugin_xcone_sub2.get());
-    ClusterSequenceArea clust_seq_sub2(particle_in_fat2, jet_def_sub2, area_def);
-    subjets_2 = sorted_by_pt(clust_seq_sub2.inclusive_jets(0));
-    for (unsigned int k = 0; k < subjets_2.size(); ++k) subjet2_area.push_back(subjets_2[k].area());
-  }
-
-  double jet1_area = 0;
-  double jet2_area = 0;
-  //double jet1_area = fatjets[0].area();
-  //double jet2_area = fatjets[1].area();
-
-  // pat-ify jets and subjets
-  auto jetCollection = std::make_unique<pat::JetCollection>();
+    // check if there are more particles then required subjets
+    bool enoughParticles = (particle_in_fatjet.size() > NSubJets_);
+    if (!enoughParticles) {
```

Comment:
I guess it makes more sense to only issue the LogWarning if `doSubjets` is true

Reply:
Infact, do we even want the warning? I find I get it frequently enough that it is a little annoying, and I don't know if it adds anything useful. As someone who actually knows/uses XCone, if you think its useful then keep it, otherwise let's remove it. Or downgrade it to a LogDebug or something

Reply:
Yes, I think we won't need the warning, I will remove this LogWarning.